dali/operators/random/noise/salt_and_pepper_noise.h

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_NOISE_SALT_AND_PEPPER_NOISE_H_
+#define DALI_OPERATORS_RANDOM_NOISE_SALT_AND_PEPPER_NOISE_H_
+
+#include <limits>
+#include "dali/operators/random/rng_base.h"
+#include "dali/operators/random/rng_base_cpu.h"
+#include "dali/operators/random/rng_base_gpu.h"
+#include "dali/pipeline/operator/arg_helper.h"
+
+#define DALI_SALT_AND_PEPPER_NOISE_TYPES \
+  uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, int64_t, float
+
+namespace dali {
+
+template <typename Backend, typename T>
+class SaltAndPepperNoiseImpl {
+ public:
+  using DistType = typename std::conditional_t<std::is_same<Backend, GPUBackend>::value,
+                                               curand_uniform_dist<float>,
+                                               std::uniform_real_distribution<float>>;
+  static constexpr T kDefaultSalt =
+      std::is_floating_point<T>::value ? T(1) : std::numeric_limits<T>::max();
+  static constexpr T kDefaultPepper =
+      std::is_floating_point<T>::value ? T(-1) : std::numeric_limits<T>::min();
+
+  DALI_HOST_DEV explicit SaltAndPepperNoiseImpl(float noise_prob = 0.05,
+                                                float salt_to_pepper_prob = 0.5,
+                                                T salt_val = kDefaultSalt,
+                                                T pepper_val = kDefaultPepper)
+      : noise_prob_(clamp<float>(noise_prob, 0, 1)),
+        salt_prob_(noise_prob_ * clamp<float>(salt_to_pepper_prob, 0, 1)),
```

Comment:
Do we want to clamp or enforce ranges in LL112.

Reply:
I'll change to enforce.

Reply:
Is this constructor ever used in device code? If not, we should probably throw `std::range_error`. If it is, clamping _here_ is the right thing to do.